src/backend/distributed/deparser/ruleutils_13.c

```diff
@@ -5290,6 +5292,27 @@ get_rule_expr(Node *node, deparse_context *context,
 				CoerceViaIO *iocoerce = (CoerceViaIO *) node;
 				Node	   *arg = (Node *) iocoerce->arg;
 
+				/*
```

Comment:
Why did we end-up doing it on the deparser? I though you were trying to fix this in the planner. There are some disadvantages of doing it in the deparser, here are my thoughts:

1) It is too late :) This is not a deparsing problem, so seems like should be fixed where the problem arise? (e.g., in the planner)
2) Relatively hard to follow, because we don't do these types of modifications during the deparse. Deparse is (should be) a read-only operation.
3) We do not deparse queries that go through standard_planner. This can be confusing addition, the reader of this code can think that we always deparse queries that passed through the standard_planner. There are obvious cases (like complex subqueries) that cannot be parsed back after standard_planner. We only need this because we use some expressions in re-written query, not the whole query.
4) Any change in deparser is a potential conflict with future Postgres versions. I'd rather avoid if there is another way to solve it. I think here we have?

So, overall, I'd be more inclined to see this (or any other change) on the SqlTaskList function instead of here.

Reply:
>I'd rather avoid if there is another way to solve it. I think here we have?

It gets more complicated to figure out where this problematic node is in more complicated queries. The deparsing logic is already visiting all nodes recursively so it does that for us. For example, the following also fails in Citus:
`SELECT cast('2008-04-07 00:00:00'::varchar as timestamp(6) without time zone) AS dummy FROM any_dist_table;`
Not sure whether the list is correct but to give an idea the node could be in
query->jointree->quals
query->targetList
query->returningList
query->jointree->fromlist
...
And for each of these lists we call get_rule_expr recursively to get to the `CoerceViaIO` node.

> Relatively hard to follow, because we don't do these types of modifications during the deparse. Deparse is (should be) a read-only operation.

As I linked here, we already have a very similar addition in this `get_rule_expr` function in Relabel types. Deparse is already not a read-only operation as I stated before, such changes for Citus are very easily done in `get_rule_expr`.

Do you still think we should do this differently?

Reply:
> query->jointree->quals
query->targetList
query->returningList
query->jointree->fromlist
...possibly more lists

All is part of "query" so we should be doing a `query_tree_mutator`, not individual conversions. `query_tree_mutator` traverses the queryTree, see `ResolveExternalParams` as an example.

> As I linked here, we already have a very similar addition in this get_rule_expr function in Relabel types. Deparse is already not a read-only operation as I stated before, such changes for Citus are very easily done in get_rule_expr.

That sounds like not the ideal thing we do. I would NOT expect to call a deparser and let it change the input. We can decide to follow not-good practice here or try to make it more intuitive. 

> Do you still think we should do this differently?

Yes, I do. We should only do this if we had to. Now, I'm even in favor of moving the Relabel type related conversion to the new `query_tree_mutator` we plan to write.

I think the follow up question is this: Do you think is it a good idea to do in the parser? I can see that this is easier, but I'm not very comfortable with such a change. I'd at least want to see a sketch of doing `query_tree_mutator` on the `SQLTaskList` and see how it looks. 

If that turns out to be too complex, we might consider getting back to this solution.

Does that work?

Reply:
> query_tree_mutator traverses the queryTree, see ResolveExternalParams as an example.

I was not aware of this, that simplifies things a lot :)

> Do you think is it a good idea to do in the parser?

If query_tree_mutator works as I am imagining right now it would work, I am all in favor of following your suggestion. The main reason I chose the deparser was that it was doing the traversing work for me.

> Now, I'm even in favor of moving the Relabel type related conversion to the new query_tree_mutator we plan to write.

If all turns out beautifully, +1 on that.

> If that turns out to be too complex, we might consider getting back to this solution. Does that work?

Yes, thanks for your suggestion. I will work on that and update the PR accordingly.